libcontainer/container_linux.go

```diff
@@ -823,23 +805,24 @@ func (c *linuxContainer) currentState() (*State, error) {
 func (c *linuxContainer) orderNamespacePaths(namespaces map[configs.NamespaceType]string, doInit bool) ([]string, error) {
 	paths := []string{}
 	nsTypes := []configs.NamespaceType{
+		configs.NEWUSER,
 		configs.NEWIPC,
 		configs.NEWUTS,
 		configs.NEWNET,
 		configs.NEWPID,
 		configs.NEWNS,
 	}
-	// For now, only join user namespace if this is an exec in process and the
-	// container supports user namespace
-	if !doInit && c.config.Namespaces.Contains(configs.NEWUSER) {
-		nsTypes = append(nsTypes, configs.NEWUSER)
-	}
```

Comment:
the order is important, especially when it comes to the user namespace. If you join the network namespace for example and then unshare the user namespace, you have no capabilities in the network namespace. You can access the network devices, but you can't bind a low port for example.

I personally think that joining the user namespace first is the best order. This is how cloneflags and unshare behaves when multiple namespaces are specified at once. We have two options:

1) specify an order that joins/unshares always happens in. The order by @dqminh + the addition of  userns makes the most sense to me.

2) allow the user to control the order. This would mean that the order of the namespace list in config.json is important and we join/unshare in that order. This would require noting in the spec that the order of namespaces matters.

I think 2 is the better option, but if we are going to specify the order, the userns should come first.

In regards to join all first then unshare all, that is fine but the order of joining still matters and I think it complicates the logic without a lot of benefit. I also suspect it may break permissions if you attempt to join an existing mount namespace and subsequently unshare the user namespace. Once again I think letting the user specify which order these things should occur by ordering the list of namespaces in the config.json is the best option here.